src/AdventureGoldGovernance.sol

```diff
@@ -0,0 +1,109 @@
+/// Adventure Gold Governance tokens in return.
+/// Users can also burn Adventure Gold Governance tokens to withdraw AGLD tokens.
+/// @dev The contract inherits from ERC20Burnable, ERC20Permit, and ERC20Votes.
+contract AdventureGoldGovernance is IERC6372, ERC20Burnable, ERC20Permit, ERC20Votes {
+    using SafeERC20 for IERC20;
+
+    /// @notice Error emitted when transfers are not to or from the contract.
+    error OnlyTransfersToFromContract();
+
+    /// @notice Emitted when a user deposits AGLD tokens, minting Adventure Gold
+    /// Governance tokens.
+    /// @param user The address of the user who deposited tokens.
+    /// @param amount The amount of AGLD tokens deposited.
+    event Deposit(address indexed user, uint256 amount);
+
+    /// @notice Emitted when a user withdraws AGLD tokens, burning Adventure
+    /// Gold Governance tokens.
+    /// @param user The address of the user who withdrew tokens.
+    /// @param amount The amount of AGLD tokens withdrawn.
+    event Withdrawal(address indexed user, uint256 amount);
+
+    /// @notice The address of the AGLD token contract.
+    address public immutable AGLD_TOKEN_ADDRESS = 0x32353A6C91143bfd6C7d363B546e62a9A2489A20;
+    /// @notice The AGLD token contract.
+    IERC20 public immutable AGLD_TOKEN = IERC20(AGLD_TOKEN_ADDRESS);
+
+    /// @notice Constructor for the Adventure Gold Governance contract.
+    constructor() ERC20("Adventure Gold Governance", "AGLDGOV") ERC20Permit("Adventure Gold Governance") {}
+
+    /// @notice Deposits AGLD tokens and mints Adventure Gold Governance tokens in return.
+    /// @param amount The amount of AGLD tokens to deposit.
+    /// @dev The AGLD tokens are transferred from the msg.sender to this contract.
+    function deposit(uint256 amount) external {
+        // Follows the Checks-Effects-Interactions pattern
+        // Checks
+        // Allowance and transfer checks occur in the AGLD token contract
+
+        // Effects
+        // Mint the same amount of Adventure Gold Governance tokens
+        _mint(msg.sender, amount);
+
+        // Interactions
+        // Transfer AGLD tokens from the sender to this contract
+        AGLD_TOKEN.safeTransferFrom(msg.sender, address(this), amount);
+
+        emit Deposit(msg.sender, amount);
+    }
+
+    /// @notice Burns Adventure Gold Governance tokens and withdraws AGLD tokens in return.
+    /// @param amount The amount of Adventure Gold Governance tokens to burn.
+    /// @dev The Adventure Gold Governance tokens are burned from the msg.sender
+    /// and the AGLD tokens are transferred to the msg.sender.
+    function withdraw(uint256 amount) external {
+        // Follows the Checks-Effects-Interactions pattern
+        // Checks
+        // Allowance checks occur in `burnFrom` via `_spendAllowance`
+
+        // Effects
+        // Burn the same amount of Adventure Gold Governance tokens
+        burnFrom(msg.sender, amount);
```

Comment:
Thoughts on using `burn` here vs `burnFrom`?

Reply:
Or even use `_burn` directly.

Reply:
I chose `burnFrom` for the allowance check, but since the tokens will always be burned for the sender, it does seem like `burn` or `_burn` is cleaner.

Even though `burn` and `_burn` are identical, I'll use `burn` so that it's easier for forks to override the burn behavior